pkg/virt-controller/watch/vm.go

```diff
@@ -231,6 +284,77 @@ func (c *VMController) orphan(cm *controller.VirtualMachineControllerRefManager,
 	return nil
 }
 
+func createDataVolumeManifest(dataVolume *cdiv1.DataVolume, vm *virtv1.VirtualMachine) *cdiv1.DataVolume {
+
+	newDataVolume := dataVolume.DeepCopy()
+
+	labels := map[string]string{}
+	annotations := map[string]string{}
+
+	labels[virtv1.CreatedByLabel] = string(vm.UID)
+	annotations[virtv1.OwnedByAnnotation] = "virt-controller"
+
+	for k, v := range dataVolume.Labels {
+		annotations[k] = v
+	}
+	for k, v := range dataVolume.Labels {
+		labels[k] = v
+	}
+	newDataVolume.ObjectMeta.Labels = labels
+	newDataVolume.ObjectMeta.Annotations = annotations
+
+	tr := true
+	newDataVolume.ObjectMeta.OwnerReferences = []v1.OwnerReference{{
+		APIVersion:         virtv1.VirtualMachineGroupVersionKind.GroupVersion().String(),
+		Kind:               virtv1.VirtualMachineGroupVersionKind.Kind,
+		Name:               vm.Name,
+		UID:                vm.UID,
+		Controller:         &tr,
+		BlockOwnerDeletion: &tr,
+	}}
+	return newDataVolume
+}
+
+func (c *VMController) handleDataVolumes(vm *virtv1.VirtualMachine, dataVolumes []*cdiv1.DataVolume) (bool, error) {
+	ready := true
+	vmKey, err := controller.KeyFunc(vm)
+	if err != nil {
+		return ready, err
+	}
+	for _, template := range vm.Spec.DataVolumeTemplates {
+		var curDataVolume *cdiv1.DataVolume
+		exists := false
+		for _, curDataVolume = range dataVolumes {
+			if curDataVolume.Name == template.Name {
+				exists = true
+				break
+			}
+		}
+		if !exists {
+			// ready = false because encountered DataVolume that is not created yet
+			ready = false
+			newDataVolume := createDataVolumeManifest(&template, vm)
+
+			c.dataVolumeExpectations.ExpectCreations(vmKey, 1)
+			curDataVolume, err = c.clientset.CdiClient().CdiV1alpha1().DataVolumes(vm.Namespace).Create(newDataVolume)
+			if err != nil {
+				c.recorder.Eventf(vm, k8score.EventTypeWarning, FailedDataVolumeCreateReason, "Error creating DataVolume %s: %v", newDataVolume.Name, err)
+				c.dataVolumeExpectations.CreationObserved(vmKey)
+				return ready, fmt.Errorf("Failed to create DataVolume: %v", err)
+			}
+			c.recorder.Eventf(vm, k8score.EventTypeNormal, SuccessfulDataVolumeCreateReason, "Created DataVolume %s", curDataVolume.Name)
+		} else if curDataVolume.Status.Phase != cdiv1.Succeeded {
+			// ready = false because encountered DataVolume that is not populated yet
+			ready = false
+			if curDataVolume.Status.Phase == cdiv1.Failed {
```

Comment:
Will the datavolume controller retry in this case? If yes, then it is strange that it has a phase of `Failed`. Shouldn't there be something like an error condition on the data volume status for as long as it will retry?

Reply:
If it is not retrying would we have to  recreate a data volume if it failed in this controller?

Reply:
The DataVolume controller will not retry to populate the DataVolume once it has reached the failed phase.

I'll add logic for the VM controller to retry the DataVolume. Otherwise, there's no recovery that can occur outside of deleting the VM object and re-posting it to the cluster.

Reply:
retry logic added.